File: src/Go/pipeline_services/core/pipeline_router.go
package core

type PipelineRouter[EventType any, ContextType any] struct {
	route func(EventType) *PipelineProvider[ContextType]
}

func NewPipelineRouter[EventType any, ContextType any](
	route func(EventType) *PipelineProvider[ContextType],
) *PipelineRouter[EventType, ContextType] {
	if route == nil {
		panic("route must not be nil")
	}
	return &PipelineRouter[EventType, ContextType]{route: route}
}

func (router *PipelineRouter[EventType, ContextType]) GetPipelineProvider(
	event EventType,
) *PipelineProvider[ContextType] {
	provider := router.route(event)
	if provider == nil {
		panic("route returned nil PipelineProvider")
	}
	return provider
}

func (router *PipelineRouter[EventType, ContextType]) Run(
	event EventType,
	context ContextType,
) ContextType {
	return router.GetPipelineProvider(event).Run(context)
}

File: src/CSharp/pipeline_services/core/PipelineRouter.cs
using System;

namespace PipelineServices.Core;

public sealed class PipelineRouter<EventType, ContextType>
{
    private readonly Func<EventType, PipelineProvider<ContextType>> route;

    public PipelineRouter(Func<EventType, PipelineProvider<ContextType>> route)
    {
        this.route = route ?? throw new ArgumentNullException(nameof(route));
    }

    public PipelineProvider<ContextType> GetPipelineProvider(EventType eventValue)
        => route(eventValue)
            ?? throw new InvalidOperationException("route returned null");

    public ContextType Run(EventType eventValue, ContextType context)
        => GetPipelineProvider(eventValue).Run(context);
}
